controllers/cloudstackfailuredomain_controller.go

```diff
@@ -130,21 +140,222 @@ func (r *CloudStackFailureDomainReconciliationRunner) ClearMachines() (ctrl.Resu
+			return result, err
+		}
+	}
+
+	result, err := triggerEtcdClusterRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	result, err = triggerControlPlaneRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	result, err = triggerMachineDeploymentRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	if len(machines.Items) > 0 {
+		return r.RequeueWithMessage("FailureDomain still has machine(s) in it.")
+	}
+	return ctrl.Result{}, nil
+}
+
+func triggerEtcdClusterRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	var etcdMachine infrav1.CloudStackMachine
+	etcdMachineFound := false
+	for _, machine := range machines {
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == "EtcdadmCluster" {
+				etcdMachineFound = true
+				etcdMachine = machine
+			}
+		}
+	}
+	if !etcdMachineFound {
+		r.Log.Info("Clear machine: no etcd VM found.")
+		return ctrl.Result{}, nil
+	}
+	for _, ref := range etcdMachine.OwnerReferences {
+		if ref.Kind == "EtcdadmCluster" {
+			r.Log.Info("Clear machine: getting etcdadmcluster ...")
+			etcdadmCluster := &unstructured.Unstructured{}
+			etcdadmCluster.SetGroupVersionKind(schema.FromAPIVersionAndKind(ref.APIVersion, ref.Kind))
+			if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: etcdMachine.Namespace, Name: ref.Name}, etcdadmCluster); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info("Clear machine: retrieved etcdadmcluster.")
+			csMachineTemplateName, ok := etcdMachine.Annotations["cluster.x-k8s.io/cloned-from-name"]
+			if !ok {
+				result, err := r.RequeueWithMessage("annotations['cluster.x-k8s.io/cloned-from-name'] not found.", "cloudstackmachine", etcdMachine.Name)
+				return result, err
+			}
+			r.Log.Info(fmt.Sprintf("Clear machine: machine template name %s decided", csMachineTemplateName))
+			csMachineTemplateNameInEtcdadmCluster, found, err := unstructured.NestedString(etcdadmCluster.Object, "spec", "infrastructureTemplate", "name")
+			if err != nil || !found {
+				return ctrl.Result{}, errors.New("etcdadmcluster spec.infrastructureTemplate.name not found or not string")
+			}
+			if csMachineTemplateName != csMachineTemplateNameInEtcdadmCluster {
+				return ctrl.Result{}, errors.Errorf("cloudstackmachinetemplate in machine %s and etcdadmcluster %s are different", etcdMachine.Name, ref.Name)
+			}
+			csMachineTemplate := &infrav1.CloudStackMachineTemplate{}
+			if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: etcdMachine.Namespace, Name: csMachineTemplateName}, csMachineTemplate); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info(fmt.Sprintf("Clear machine: machine template name %s retrieved", csMachineTemplateName))
+			csMachineTemplate.Name = fmt.Sprintf("%s-template-%d", ref.Name, time.Now().UnixNano()/int64(time.Millisecond))
+			csMachineTemplate.SetAnnotations(map[string]string{})
```

Comment:
can you explain how you're triggering the rollout? It seems like you're creating a new cloudstackmachinetemplate and then setting that on the etcdadmcluster. But won't that mess up EKS-A's controller, which is generating and applying its own CloudStackMachineTemplates and EtcdAdmCluster resources?

Reply:
when etcdadmcluster's infrastructureTemplate cloudstack template get changed, a rollout will be triggered for etcdadmcluster. 

The initial settings are generated by eksa, but changing its settings later on seems fine - when eksa upgrade, it will generate a new cloudstack machine template.

Reply:
Curious... I too share your concern here @maxdrib.

How exactly is there no conflict? Does EKS-A somehow get feedback on this?

I would think modifying the (input) specs on most any of the CAPI/CAPC custom resources would be fought by EKS-A's incessant application of its spec. 

I like using incessant here. haha : p